nest-forms-backend/src/utils/subservices/sharepoint.subservice.ts

```diff
@@ -0,0 +1,357 @@
+import { OnQueueFailed, Process, Processor } from '@nestjs/bull'
+import { Injectable, Logger } from '@nestjs/common'
+import { FormError, Forms, FormState } from '@prisma/client'
+import { GenericObjectType } from '@rjsf/utils'
+import axios, { AxiosResponse } from 'axios'
+import { Job } from 'bull'
+import {
+  FormDefinition,
+  FormDefinitionType,
+} from 'forms-shared/definitions/formDefinitionTypes'
+import { getFormDefinitionBySlug } from 'forms-shared/definitions/getFormDefinitionBySlug'
+import { omitExtraData } from 'forms-shared/form-utils/omitExtraData'
+import {
+  getArrayForOneToMany,
+  getValuesForFields,
+} from 'forms-shared/sharepoint/getValuesForSharepoint'
+import { escape, filter, List } from 'lodash'
+
+import {
+  FormsErrorsEnum,
+  FormsErrorsResponseEnum,
+} from '../../forms/forms.errors.enum'
+import PrismaService from '../../prisma/prisma.service'
+import ThrowerErrorGuard from '../guards/thrower-error.guard'
+import {
+  getFrontendFormTitleFromForm,
+  getSubjectTextFromForm,
+} from '../handlers/text.handler'
+import alertError from '../logging'
+import {
+  SharepointErrorsEnum,
+  SharepointErrorsResponseEnum,
+} from './dtos/sharepoint.errors.enum'
+
+@Injectable()
+@Processor('sharepoint')
+export default class SharepointSubservice {
+  private readonly logger: Logger
+
+  constructor(
+    private throwerErrorGuard: ThrowerErrorGuard,
+    private prismaService: PrismaService,
+  ) {
+    this.logger = new Logger('SharepointSubservice')
+
+    if (
+      !process.env.SHAREPOINT_TENANT_ID ||
+      !process.env.SHAREPOINT_CLIENT_ID ||
+      !process.env.SHAREPOINT_CLIENT_SECRET ||
+      !process.env.SHAREPOINT_DOMAIN ||
+      !process.env.SHAREPOINT_URL
+    ) {
+      throw new Error(
+        'Missing Sharepoint .env values, one of: SHAREPOINT_TENANT_ID, SHAREPOINT_CLIENT_ID, SHAREPOINT_CLIENT_SECRET, SHAREPOINT_DOMAIN, SHAREPOINT_URL',
+      )
+    }
+  }
+
+  @Process()
+  async transcode(job: Job<{ formId: string }>): Promise<void> {
+    await this.postNewRecord(job.data.formId)
+  }
+
+  @OnQueueFailed()
+  handler(job: Job<{ formId: string }>, err: Error): void {
+    alertError(
+      `Sending form ${job.data.formId} to Sharepoint has failed. Number of tries for this form so far: ${job.attemptsMade}`,
+      this.logger,
+      JSON.stringify(err),
+    )
+
+    this.prismaService.forms
+      .update({
+        where: {
+          id: job.data.formId,
+        },
+        data: {
+          error: FormError.SHAREPOINT_SEND_ERROR,
+        },
+      })
+      .catch((error) => {
+        alertError(
+          `Setting form error with id ${job.data.formId} to POWERAPPS_SEND_ERROR failed.`,
+          this.logger,
+          JSON.stringify(error),
+        )
+      })
+  }
+
+  public async postNewRecord(formId: string): Promise<void> {
```

Comment:
There are many `postDataToSharepoint` calls in this function, what if some of them succeed and some of them fail? Isn't there gonna be a mess in Sharepoint?

Reply:
Yes, but I don't see a solution for this. We alert an error when this happens, so we can somehow resend it to sharepoint. Or we can retry failed calls in some queue?

Reply:
Yes, we should definitely have some retry mechanism, however we can still catch and alert the errors, and in the worst case just resend the forms to sharepoint.

Currently it is an add operation, if there is some record, it will add the same. I suppose we can change it to rewrite, but we would need a unique key for each record (even for the oneToMany records, like children in this case). I am creating an issue for this.

Reply:
https://github.com/bratislava/konto.bratislava.sk/issues/1243